src/gbm/gbtree.cc

```diff
@@ -475,14 +509,277 @@ class GBTree : public GradientBooster {
   std::vector<std::unique_ptr<TreeUpdater> > updaters;
 };
 
+// dart
+class Dart : public GBTree {
+ public:
+  Dart() {
+    num_pbuffer = 0;
+    weight_drop.clear();
+    idx_drop.clear();
+  }
+
+  void Configure(const std::vector<std::pair<std::string, std::string> >& cfg) override {
+    GBTree::Configure(cfg);
+    if (trees.size() == 0) {
+      dparam.InitAllowUnknown(cfg);
+    }
+  }
+
+  void Load(dmlc::Stream* fi) override {
+    GBTree::Load(fi);
+    weight_drop.resize(mparam.num_trees);
+    if (mparam.num_trees != 0) {
+      CHECK_EQ(fi->Read(dmlc::BeginPtr(weight_drop), sizeof(float) * mparam.num_trees),
+               sizeof(float) * mparam.num_trees)
+          << "Dart: invalid model file";;
+    }
+  }
+
+  void Save(dmlc::Stream* fo) const override {
+    GBTree::Save(fo);
+    if (weight_drop.size() != 0) {
+      fo->Write(dmlc::BeginPtr(weight_drop), sizeof(float) * tree_info.size());
+    }
+  }
+
+  void DoBoost(DMatrix* p_fmat,
+               int64_t buffer_offset,
+               std::vector<bst_gpair>* in_gpair) override {
+    const std::vector<bst_gpair>& gpair = *in_gpair;
```

Comment:
Is this function different from parent's DoBoost? If only difference is CcommitModel, we can make CommitModel virtual and override that function